packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -200,13 +299,15 @@ export class TokenDetectionController extends StaticIntervalPollingController<
         this.#networkClientId = selectedNetworkClientId;
         const newChainId = this.#getCorrectChainId(selectedNetworkClientId);
         const isChainIdChanged = this.#chainId !== newChainId;
-        this.#chainId = newChainId;
 
         this.#isDetectionEnabledForNetwork =
           isTokenDetectionSupportedForNetwork(newChainId);
 
-        if (this.#isDetectionEnabledForNetwork && isChainIdChanged) {
-          await this.detectTokens();
+        if (isChainIdChanged && this.#isDetectionEnabledForNetwork) {
+          this.#chainId = newChainId;
+          await this.#restartTokenDetection({
+            chainId: this.#chainId,
```

Comment:
Since we will need the network client ID later, can we pass it instead of chain ID?

``` suggestion
            networkClientId: this.#networkClientId,

Reply:
Fixed here: https://github.com/MetaMask/core/pull/3775/commits/e40e493d3aec680ad5b83f65885b25e4fe7b7b06